File: src/rust/fab/src/preflight.rs
use anyhow::{bail, Result};
use std::fs;
use std::path::PathBuf;

use crate::resolve;
use crate::status;
use crate::statusfile::{self, Checklist, Confidence, StageState};

/// Result holds the structured preflight output.
pub struct PreflightResult {
    pub id: String,
    pub name: String,
    pub change_dir: String,
    pub stage: String,
    pub display_stage: String,
    pub display_state: String,
    pub progress: Vec<StageState>,
    pub checklist: Checklist,
    pub confidence: Confidence,
}

/// Performs preflight validation and returns structured result.
pub fn run(fab_root: &str, change_override: &str) -> Result<PreflightResult> {
    // 1. Check project initialization
    let config_path = PathBuf::from(fab_root).join("project").join("config.yaml");
    if !config_path.exists() {
        bail!("Project not initialized — fab/project/config.yaml not found. Run /fab-setup.");
    }
    let const_path = PathBuf::from(fab_root)
        .join("project")
        .join("constitution.md");
    if !const_path.exists() {
        bail!("Project not initialized — fab/project/constitution.md not found. Run /fab-setup.");
    }

    // 2. Sync staleness warning (non-blocking)
    check_sync_staleness(fab_root);

    // 3. Resolve change
    let folder = resolve::to_folder(fab_root, change_override)?;

    // 4. Check change directory
    let change_dir = PathBuf::from(fab_root).join("changes").join(&folder);
    if !change_dir.exists() {
        bail!("Change directory not found: fab/changes/{}", folder);
    }

    // 5. Check .status.yaml
    let status_path = change_dir.join(".status.yaml");
    if !status_path.exists() {
        bail!(".status.yaml not found in fab/changes/{}", folder);
    }

    // 6. Load and validate
    let sf = statusfile::load(&status_path.to_string_lossy())
        .map_err(|e| anyhow::anyhow!("Failed to load .status.yaml: {}", e))?;

    status::validate(&sf).map_err(|e| anyhow::anyhow!("Invalid .status.yaml: {}", e))?;

    // Build result
    let id = resolve::extract_id(&folder);
    let current_stage = status::current_stage(&sf);
    let (display_stage, display_state) = status::display_stage(&sf);
    let rel_change_dir = format!("fab/changes/{}", folder);

    Ok(PreflightResult {
        id,
        name: folder,
        change_dir: rel_change_dir,
        stage: current_stage,
        display_stage,
        display_state,
        progress: sf.get_progress_map(),
        checklist: sf.checklist,
        confidence: sf.confidence,
    })
}

/// Produces the YAML output matching preflight.sh format.
pub fn format_yaml(r: &PreflightResult) -> String {
    let mut b = String::new();

    b.push_str(&format!("id: {}\n", r.id));
    b.push_str(&format!("name: {}\n", r.name));
    b.push_str(&format!("change_dir: {}\n", r.change_dir));
    b.push_str(&format!("stage: {}\n", r.stage));
    b.push_str(&format!("display_stage: {}\n", r.display_stage));
    b.push_str(&format!("display_state: {}\n", r.display_state));
    b.push_str("progress:\n");
    for ss in &r.progress {
        b.push_str(&format!("  {}: {}\n", ss.stage, ss.state));
    }
    b.push_str("checklist:\n");
    b.push_str(&format!("  generated: {}\n", r.checklist.generated));
    b.push_str(&format!("  completed: {}\n", r.checklist.completed));
    b.push_str(&format!("  total: {}\n", r.checklist.total));
    b.push_str("confidence:\n");
    b.push_str(&format!("  certain: {}\n", r.confidence.certain));
    b.push_str(&format!("  confident: {}\n", r.confidence.confident));
    b.push_str(&format!("  tentative: {}\n", r.confidence.tentative));
    b.push_str(&format!("  unresolved: {}\n", r.confidence.unresolved));
    b.push_str(&format!("  score: {:.1}\n", r.confidence.score));

    if let Some(true) = r.confidence.indicative {
        b.push_str("  indicative: true\n");
    }

    b
}

fn check_sync_staleness(fab_root: &str) {
    let version_file = PathBuf::from(fab_root).join(".kit").join("VERSION");
    let kit_version = fs::read_to_string(&version_file)
        .ok()
        .map(|s| s.trim().to_string())
        .unwrap_or_default();

    let sync_file = PathBuf::from(fab_root).join(".kit-sync-version");
    match fs::read_to_string(&sync_file) {
        Ok(data) => {
            let sync_version = data.trim().to_string();
            if !kit_version.is_empty() && !sync_version.is_empty() && kit_version != sync_version {
                eprintln!(
                    "\u{26a0} Skills out of sync — run fab-sync.sh to refresh (engine {}, last synced {})",
                    kit_version, sync_version
                );
            }
        }
        Err(_) => {
            if !kit_version.is_empty() {
                eprintln!(
                    "\u{26a0} Skills may be out of sync — run fab-sync.sh to refresh"
                );
            }
        }
    }
}

File: src/go/fab/internal/preflight/preflight.go
package preflight

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/sahil87/fab-kit/src/go/fab/internal/resolve"

	"github.com/sahil87/fab-kit/src/go/fab/internal/status"
	sf "github.com/sahil87/fab-kit/src/go/fab/internal/statusfile"
	"gopkg.in/yaml.v3"
)

// Result holds the structured preflight output.
type Result struct {
	ID           string
	Name         string
	ChangeDir    string
	Stage        string
	DisplayStage string
	DisplayState string
	Progress     []sf.StageState
	Checklist    sf.Checklist
	Confidence   sf.Confidence
}

// Run performs preflight validation and returns structured result.
func Run(fabRoot, changeOverride string) (*Result, error) {
	// 1. Check project initialization
	configPath := filepath.Join(fabRoot, "project", "config.yaml")
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		return nil, fmt.Errorf("Project not initialized — fab/project/config.yaml not found. Run /fab-setup.")
	}
	constPath := filepath.Join(fabRoot, "project", "constitution.md")
	if _, err := os.Stat(constPath); os.IsNotExist(err) {
		return nil, fmt.Errorf("Project not initialized — fab/project/constitution.md not found. Run /fab-setup.")
	}

	// 2. Sync staleness warning (non-blocking)
	checkSyncStaleness(fabRoot)

	// 3. Resolve change
	folder, err := resolve.ToFolder(fabRoot, changeOverride)
	if err != nil {
		return nil, err
	}

	// 4. Check change directory
	changeDir := filepath.Join(fabRoot, "changes", folder)
	if _, err := os.Stat(changeDir); os.IsNotExist(err) {
		return nil, fmt.Errorf("Change directory not found: fab/changes/%s", folder)
	}

	// 5. Check .status.yaml
	statusPath := filepath.Join(changeDir, ".status.yaml")
	if _, err := os.Stat(statusPath); os.IsNotExist(err) {
		return nil, fmt.Errorf(".status.yaml not found in fab/changes/%s", folder)
	}

	// 6. Load and validate
	statusFile, err := sf.Load(statusPath)
	if err != nil {
		return nil, fmt.Errorf("Failed to load .status.yaml: %w", err)
	}

	if err := status.Validate(statusFile); err != nil {
		return nil, fmt.Errorf("Invalid .status.yaml: %w", err)
	}

	// Build result
	id := resolve.ExtractID(folder)
	currentStage := status.CurrentStage(statusFile)
	displayStage, displayState := status.DisplayStage(statusFile)

	relChangeDir := "fab/changes/" + folder

	return &Result{
		ID:           id,
		Name:         folder,
		ChangeDir:    relChangeDir,
		Stage:        currentStage,
		DisplayStage: displayStage,
		DisplayState: displayState,
		Progress:     statusFile.GetProgressMap(),
		Checklist:    statusFile.Checklist,
		Confidence:   statusFile.Confidence,
	}, nil
}

// FormatYAML produces the YAML output matching preflight.sh format.
func FormatYAML(r *Result) string {
	var b strings.Builder

	fmt.Fprintf(&b, "id: %s\n", r.ID)
	fmt.Fprintf(&b, "name: %s\n", r.Name)
	fmt.Fprintf(&b, "change_dir: %s\n", r.ChangeDir)
	fmt.Fprintf(&b, "stage: %s\n", r.Stage)
	fmt.Fprintf(&b, "display_stage: %s\n", r.DisplayStage)
	fmt.Fprintf(&b, "display_state: %s\n", r.DisplayState)
	b.WriteString("progress:\n")
	for _, ss := range r.Progress {
		fmt.Fprintf(&b, "  %s: %s\n", ss.Stage, ss.State)
	}
	b.WriteString("checklist:\n")
	fmt.Fprintf(&b, "  generated: %v\n", r.Checklist.Generated)
	fmt.Fprintf(&b, "  completed: %d\n", r.Checklist.Completed)
	fmt.Fprintf(&b, "  total: %d\n", r.Checklist.Total)
	b.WriteString("confidence:\n")
	fmt.Fprintf(&b, "  certain: %d\n", r.Confidence.Certain)
	fmt.Fprintf(&b, "  confident: %d\n", r.Confidence.Confident)
	fmt.Fprintf(&b, "  tentative: %d\n", r.Confidence.Tentative)
	fmt.Fprintf(&b, "  unresolved: %d\n", r.Confidence.Unresolved)
	fmt.Fprintf(&b, "  score: %.1f\n", r.Confidence.Score)

	if r.Confidence.Indicative != nil && *r.Confidence.Indicative {
		b.WriteString("  indicative: true\n")
	}

	return b.String()
}

func checkSyncStaleness(fabRoot string) {
	kitVersion := ""
	versionFile := filepath.Join(fabRoot, ".kit", "VERSION")
	if data, err := os.ReadFile(versionFile); err == nil {
		kitVersion = strings.TrimSpace(string(data))
	}
	if kitVersion == "" {
		return
	}

	configPath := filepath.Join(fabRoot, "project", "config.yaml")
	configData, err := os.ReadFile(configPath)
	if err != nil {
		return
	}

	var cfg struct {
		FabVersion string `yaml:"fab_version"`
	}
	if err := yaml.Unmarshal(configData, &cfg); err != nil {
		return
	}
	configVersion := strings.TrimSpace(cfg.FabVersion)
	if configVersion == "" {
		return
	}

	if kitVersion != configVersion {
		fmt.Fprintf(os.Stderr, "⚠ Skills may be out of sync — run fab sync to refresh (engine %s, project %s)\n", kitVersion, configVersion)
	}
}
